pyteomics/proforma.py

```diff
@@ -2534,3 +2563,56 @@ def find_tags_by_id(self, tag_id, include_position=True):
     @property
     def tags(self):
         return [tag for tags_at in [pos[1] for pos in self if pos[1]] for tag in tags_at]
+
+    def composition(self, include_charge=False, aa_comp=None):
+        '''
+        Calculate the elemental composition of the ProForma sequence.
+
+        Parameters
+        ----------
+        include_charge : bool, optional
+            If True, the charge state will be included in the composition.
+            Otherwise, composition of the neutral molecule will be returned.
+            Defaults to False.
+        aa_comp : dict, optional
+            A dictionary mapping amino acid symbols to their respective
+            compositions. If not provided, the standard amino acid composition
+            will be used.
+
+        Returns
+        -------
+        Composition
+            :py:class:`Composition` object representing the composition of the ProForma sequence.
+        '''
+        comp = Composition()
+        if aa_comp is None:
+            aa_comp = std_aa_comp
+        try:
+            for aa, tags in self.sequence:
+                comp += aa_comp[aa]
+                for tag in tags or []:
+                    comp += tag.composition
+            for tag in chain(self.fixed_modifications, self.labile_modifications, self.unlocalized_modifications):
+                comp += tag.composition
```

Comment:
`tag.composition` may be `None` if you have a mass modification like `[+79.01]`. What to do in this case is kind of hard to answer, which is part of why I left it out this method. We could probably parameterize this like `error_on_missing_composition: bool = True` and let the user control whether this is an error or not.

Reply:
Oh sure. I forgot that type didn't even stub out the attribute for duck typing.

All the CV-based modification tags implement `composition` as a wrapper around `self.definition.get("composition")` which would return `None` if the underlying dictionary doesn't have the key set, or if the key's value is `None`.